main/usb_uart_bridge_main.c

```diff
@@ -288,6 +244,32 @@ static void tinyusb_cdc_line_coding_changed_callback(int itf, cdcacm_event_t *ev
     uint8_t data_bits = event->line_coding_changed_data.p_line_coding->data_bits;
     ESP_LOGV(TAG, "host require bit_rate=%" PRIu32 " stop_bits=%u parity=%u data_bits=%u", bit_rate, stop_bits, parity, data_bits);
 
+    static uint8_t magic_bootloader_trigger_stage = 0;
+    static uint64_t magic_bootloader_trigger_time = 0;
+
+    uint64_t now = esp_timer_get_time();
+
+    if ((now - magic_bootloader_trigger_time > MAGIC_BOOTLOADER_TRIGGER_TIMEOUT_MICROS) && magic_bootloader_trigger_stage > 0) {
+        ESP_LOGW(TAG, "Magic bootloader baudrate period timed out, resetting state");
+        magic_bootloader_trigger_stage = 0;
+    }
+
+    if (bit_rate == MAGIC_BOOTLOADER_TRIGGER_BAUDRATES[magic_bootloader_trigger_stage]) {
+        if (magic_bootloader_trigger_stage == 0) {
+            magic_bootloader_trigger_time = now;
+        }
+
+        ESP_LOGW(TAG, "Received magic bootloader baudrate, stage %d (%" PRIu32 " baud)", magic_bootloader_trigger_stage, bit_rate);
+        magic_bootloader_trigger_stage++;
+
+        if (magic_bootloader_trigger_stage == sizeof(MAGIC_BOOTLOADER_TRIGGER_BAUDRATES) / sizeof(MAGIC_BOOTLOADER_TRIGGER_BAUDRATES[0])) {
+            ESP_LOGW(TAG, "Triggering bootloader");
+            REG_WRITE(RTC_CNTL_OPTION1_REG, RTC_CNTL_FORCE_DOWNLOAD_BOOT);
+            esp_restart();
```

Comment:
This for me is not 100% reliable for some reason. Is there a better way to do this? Maybe cleanly shut down the USB stack first and do this from a separate task?

Reply:
Did you figure this out?

Reply:
I'll test it out again later, I was focusing on the other communication PR the past couple of days.

I think breaking this out into a separate task, triggering the task from the tinyUSB callback, and then having the task tear things down cleanly before resetting may be way forward. I think I was seeing macOS not react properly to the USB device "swap" when it's done crudely.

Reply:
Moving to a separate task with a sempahore works. 20/20 resets in a row are now successful.